File: android/GreenReadAR/app/src/main/kotlin/com/greenreadar/physics/BallPhysicsEngine.kt
// BallPhysicsEngine.kt
// GreenRead AR — Golf Ball Physics Simulation (Android)
// Port of BallPhysicsEngine.swift — PRD §2.2
//
// ARCHITECTURE NOTE — Same as iOS:
// This engine uses custom Verlet integration for break prediction.
// It depends ONLY on GreenReadSession interface (not ARCoreSession).
// Same engine, same math, same results on both platforms.

package com.greenreadar.physics

import com.greenreadar.ar.Float2
import com.greenreadar.ar.Float3
import com.greenreadar.ar.GreenReadSession
import com.greenreadar.ar.SlopeAnalyzer
import com.greenreadar.ar.BallRollResult
import kotlin.math.min
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Simulates golf ball rolling on a scanned green mesh.
 * Uses custom Verlet integration for prediction accuracy.
 *
 * Port of BallPhysicsEngine.swift — identical physics.
 * terrain reference is GreenReadSession (not ARCoreSession).
 */
class BallPhysicsEngine(
    private val terrain: GreenReadSession,
    stimpmeter: Float
) {
    companion object {
        const val BALL_MASS: Float = 0.046f       // kg — standard golf ball
        const val BALL_RADIUS: Float = 0.02135f   // meters
        const val GRAVITY: Float = 9.81f          // m/s²
        const val MIN_SPEED: Float = 0.02f        // m/s — stop threshold (PRD §2.2)
    }

    // Stimpmeter-derived friction
    // PRD: Higher stimp → lower friction → faster green
    // mu ≈ 1.0 / (stimp * 0.38) — calibrated for realistic roll
    private val stimpValue: Float = stimpmeter.coerceIn(7f, 13f)
    private val frictionCoeff: Float = 1.0f / (stimpValue * 0.38f)

    // State
    var position: Float3 = Float3.zero
        private set
    var velocity: Float3 = Float3.zero
        private set
    var isRolling: Boolean = false
        private set

    // Trail recording
    val trail: MutableList<Float3> = mutableListOf()
    private var startPosition: Float3 = Float3.zero
    private var totalDistance: Float = 0f
    private val maxTrailPoints = 500

    /**
     * Initialize a roll from ball position toward hole.
     * Calculates initial speed based on distance, slope, and green speed.
     */
    fun initRoll(from ballPos: Float3, toward holePos: Float3) {
        position = ballPos
        startPosition = ballPos
        trail.clear()
        trail.add(ballPos)
        isRolling = true
        totalDistance = 0f

        // Direction toward hole
        val direction = (holePos - ballPos).normalized()
        val distance = (holePos - ballPos).length()

        // Calculate initial speed based on distance and terrain
        var speed = distance * 1.4f + 0.5f

        // Adjust for average slope along putt line
        val midPoint = Float2(
            (ballPos.x + holePos.x) / 2f,
            (ballPos.z + holePos.z) / 2f
        )

        terrain.slopeAt(midPoint)?.let { slope ->
            val slopeComponent = Float3.dot(slope.normal, direction)
            if (slopeComponent < 0) {
                // Uphill — need more speed
                speed *= 1f + kotlin.math.abs(slopeComponent) * 2.5f
            } else {
                // Downhill — need less speed
                speed *= 1f - slopeComponent * 1.2f
            }
        }

        // Adjust for green speed
        speed *= (10.0f / stimpValue)

        velocity = Float3(
            direction.x * speed,
            0f,
            direction.z * speed
        )
    }

    /**
     * Physics update — Verlet integration with substeps.
     * Identical to Swift BallPhysicsEngine.update().
     */
    data class UpdateResult(
        val position: Float3,
        val velocity: Float3,
        val isRolling: Boolean
    )

    fun update(deltaTime: Float): UpdateResult {
        if (!isRolling) {
            return UpdateResult(position, velocity, false)
        }

        val dt = min(deltaTime, 0.016f) // Cap at ~60fps
        val substeps = 4
        val subDt = dt / substeps.toFloat()

        for (step in 0 until substeps) {
            // Get slope at current position
            val pos2D = Float2(position.x, position.z)

            val slope = terrain.slopeAt(pos2D)
            if (slope == null) {
                isRolling = false
                break
            }

            // Rolling friction (opposes velocity)
            val speed = velocity.length()
            var frictionForce = Float3.zero
            if (speed > 0.001f) {
                val velNorm = velocity.normalized()
                frictionForce = velNorm * (-frictionCoeff * GRAVITY)
            }

            // Gravity along slope surface — creates the break
            val slopeAngleRad = slope.degrees * (Math.PI.toFloat() / 180f)
            val slopeGravity = slope.fallLine *
                kotlin.math.sin(slopeAngleRad) * GRAVITY

            // Total acceleration
            val accel = slopeGravity + frictionForce

            // Verlet integration
            velocity = velocity + accel * subDt
            position = position + velocity * subDt

            // Snap to terrain surface
            terrain.heightAt(pos2D)?.let { height ->
                position = Float3(position.x, height + BALL_RADIUS, position.z)
            }

            totalDistance += speed * subDt

            // Stop condition (PRD §2.2: velocity < 0.02 m/s)
            if (speed < MIN_SPEED && totalDistance > 0.1f) {
                isRolling = false
                break
            }
        }

        // Record trail
        trail.add(position)
        if (trail.size > maxTrailPoints) {
            trail.removeAt(0)
        }

        return UpdateResult(position, velocity, isRolling)
    }

    /**
     * Calculate final result after ball stops.
     */
    fun calculateResult(holePosition: Float3): BallRollResult {
        val distance = (holePosition - startPosition).length()

        // Break calculation
        val (breakAmount, breakDir) = SlopeAnalyzer.calculateBreak(
            trail = trail,
            startPosition = startPosition,
            holePosition = holePosition
        )

        // Elevation change
        val elevationChange = holePosition.y - startPosition.y

        // Average slope along path
        var totalSlope = 0f
        var slopeCount = 0f
        for (point in trail) {
            terrain.slopeAt(Float2(point.x, point.z))?.let { slope ->
                totalSlope += slope.degrees
                slopeCount += 1f
            }
        }
        val avgSlope = if (slopeCount > 0) totalSlope / slopeCount else 0f

        return BallRollResult(
            finalPosition = position,
            trailPositions = trail.toList(),
            totalDistance = totalDistance,
            maxBreak = breakAmount,
            breakDirection = breakDir,
            avgSlopeAlongPath = avgSlope,
            elevationChange = elevationChange
        )
    }
}

File: swift/GreenReadAR/AR/BallPhysicsEngine.swift
// BallPhysicsEngine.swift
// GreenRead AR — Golf Ball Physics Simulation
// PRD §2.2: Rullande bollar
// Mass: 0.046 kg, variable friction per Stimpmeter (7–13)
//
// ARCHITECTURE NOTE — Verlet vs RealityKit Physics:
// ──────────────────────────────────────────────────
// This engine uses CUSTOM Verlet integration intentionally, NOT as a replacement
// for RealityKit's PhysicsBodyComponent. The two systems work together:
//
// 1. BallPhysicsEngine (this file) — PREDICTION & BREAK CALCULATION
//    - Runs the ball simulation using Verlet integration
//    - Uses terrain slope normals to calculate gravity-driven break curves
//    - Produces: trail positions, break amount/direction, speed recommendations
//    - Reason: RealityKit's built-in physics doesn't expose enough control over
//      rolling friction coefficients tied to Stimpmeter values, nor does it
//      provide trail/path recording needed for break curve visualization
//
// 2. RealityKit PhysicsBodyComponent (in ARViewSession.swift) — COLLISION & RENDERING
//    - Applied to ball ModelEntity for mesh collision response
//    - Handles visual rolling (rotation) against the scanned mesh
//    - Provides realistic bounce/settle behavior at hole
//    - The ball entity position is SET by this engine each frame
//
// This hybrid approach gives us:
//   ✅ Precise Stimpmeter-calibrated friction (custom)
//   ✅ Accurate break prediction with trail recording (custom)
//   ✅ Realistic mesh collision and visual response (RealityKit)
//   ✅ Same engine works on both iPhone and visionOS (platform-agnostic math)
//
// PRD §10.2-B compliance: This engine depends only on the GreenReadSession
// protocol's heightAt/slopeAt methods, so it works identically on ARView
// (iPhone) and ImmersiveSpace (visionOS).

import simd
import Foundation

/// Simulates golf ball rolling on a scanned green mesh.
///
/// Uses custom Verlet integration for prediction accuracy while RealityKit
/// PhysicsBodyComponent handles collision rendering. See architecture note above.
final class BallPhysicsEngine {
    
    // MARK: - Constants
    static let ballMass: Float = 0.046       // kg — standard golf ball
    static let ballRadius: Float = 0.02135   // meters
    static let gravity: Float = 9.81         // m/s²
    static let minSpeed: Float = 0.02        // m/s — stop threshold (PRD §2.2)
    
    // MARK: - Properties
    private weak var terrain: ARViewSession?
    private let stimpmeter: Float
    private let frictionCoeff: Float
    
    // State
    var position: SIMD3<Float> = .zero
    var velocity: SIMD3<Float> = .zero
    var isRolling: Bool = false
    
    // Trail recording
    private(set) var trail: [SIMD3<Float>] = []
    private var startPosition: SIMD3<Float> = .zero
    private var totalDistance: Float = 0
    private let maxTrailPoints = 500
    
    // MARK: - Init
    init(terrain: ARViewSession, stimpmeter: Float) {
        self.terrain = terrain
        self.stimpmeter = max(7, min(13, stimpmeter))
        
        // PRD: Higher stimp → lower friction → faster green
        // mu ≈ 1 / (stimp * 0.38) — calibrated for realistic roll
        self.frictionCoeff = 1.0 / (self.stimpmeter * 0.38)
    }
    
    // MARK: - Init Roll
    func initRoll(from ballPos: SIMD3<Float>, toward holePos: SIMD3<Float>) {
        position = ballPos
        startPosition = ballPos
        trail = [ballPos]
        isRolling = true
        totalDistance = 0
        
        // Direction toward hole
        let direction = normalize(holePos - ballPos)
        let distance = length(holePos - ballPos)
        
        // Calculate initial speed based on distance and terrain
        var speed = distance * 1.4 + 0.5
        
        // Adjust for average slope along putt line
        let midPoint = SIMD2<Float>(
            (ballPos.x + holePos.x) / 2,
            (ballPos.z + holePos.z) / 2
        )
        
        if let slope = terrain?.slopeAt(position: midPoint) {
            let slopeComponent = dot(slope.normal, direction)
            if slopeComponent < 0 {
                // Uphill — need more speed
                speed *= 1 + abs(slopeComponent) * 2.5
            } else {
                // Downhill — need less speed
                speed *= 1 - slopeComponent * 1.2
            }
        }
        
        // Adjust for green speed
        speed *= (10.0 / stimpmeter)
        
        velocity = SIMD3<Float>(
            direction.x * speed,
            0,
            direction.z * speed
        )
    }
    
    // MARK: - Physics Update
    struct UpdateResult {
        let position: SIMD3<Float>
        let velocity: SIMD3<Float>
        let isRolling: Bool
    }
    
    func update(deltaTime: Float) -> UpdateResult {
        guard isRolling else {
            return UpdateResult(position: position, velocity: velocity, isRolling: false)
        }
        
        let dt = min(deltaTime, 0.016) // Cap at ~60fps
        let substeps = 4
        let subDt = dt / Float(substeps)
        
        for _ in 0..<substeps {
            // Get slope at current position
            let pos2D = SIMD2<Float>(position.x, position.z)
            
            guard let slope = terrain?.slopeAt(position: pos2D) else {
                isRolling = false
                break
            }
            
            // Rolling friction (opposes velocity)
            let speed = length(velocity)
            var frictionForce = SIMD3<Float>.zero
            if speed > 0.001 {
                frictionForce = normalize(velocity) * (-frictionCoeff * Self.gravity)
            }
            
            // Gravity along slope surface — creates the break
            let slopeGravity = slope.fallLine *
                sin(slope.degrees * .pi / 180) * Self.gravity
            
            // Total acceleration
            let accel = slopeGravity + frictionForce
            
            // Verlet integration
            velocity += accel * subDt
            position += velocity * subDt
            
            // Snap to terrain surface
            if let height = terrain?.heightAt(position: pos2D) {
                position.y = height + Self.ballRadius
            }
            
            totalDistance += speed * subDt
            
            // Stop condition (PRD §2.2: velocity < 0.02 m/s)
            if speed < Self.minSpeed && totalDistance > 0.1 {
                isRolling = false
                break
            }
        }
        
        // Record trail
        trail.append(position)
        if trail.count > maxTrailPoints {
            trail.removeFirst()
        }
        
        return UpdateResult(position: position, velocity: velocity, isRolling: isRolling)
    }
    
    // MARK: - Calculate Result
    func calculateResult(holePosition: SIMD3<Float>) -> BallRollResult {
        let distance = length(holePosition - startPosition)
        
        // Break calculation
        let breakInfo = SlopeAnalyzer.calculateBreak(
            trail: trail,
            startPosition: startPosition,
            holePosition: holePosition
        )
        
        // Elevation change
        let elevationChange = holePosition.y - startPosition.y
        
        // Average slope along path
        var totalSlope: Float = 0
        var slopeCount: Float = 0
        for point in trail {
            if let slope = terrain?.slopeAt(position: SIMD2<Float>(point.x, point.z)) {
                totalSlope += slope.degrees
                slopeCount += 1
            }
        }
        let avgSlope = slopeCount > 0 ? totalSlope / slopeCount : 0
        
        return BallRollResult(
            finalPosition: position,
            trailPositions: trail,
            totalDistance: totalDistance,
            maxBreak: breakInfo.amount,
            breakDirection: breakInfo.direction,
            avgSlopeAlongPath: avgSlope,
            elevationChange: elevationChange
        )
    }
}
